deployment/distribution/src/main/kotlin/org/codice/ctk/TestRunner.kt

```diff
@@ -49,51 +45,23 @@ internal class TestRunner {
 
         val summaryGeneratingListener = SummaryGeneratingListener()
         LauncherFactory.create().apply {
-            registerTestExecutionListeners(summaryGeneratingListener, TestNameListener())
+            registerTestExecutionListeners(summaryGeneratingListener)
         }.execute(request)
 
-        PrintWriter(System.out).use { printer ->
+        System.out.apply {
+            println()
+            println("----------------------------------")
+            println("SAML Conformance Test Kit Starting")
+            println("----------------------------------")
 
-            summaryGeneratingListener.summary.printFailuresTo(printer)
-            summaryGeneratingListener.summary.printTo(printer)
+            Report.printReport()
 
-            if (summaryGeneratingListener.summary.totalFailureCount > 0) {
-                System.out.println(Ansi.ansi().fgRed().a("TESTS FAILED").reset())
+            if (Report.hasErrors()) {
+                println(Ansi.ansi().fgRed().a("TESTS FAILED").reset())
                 System.exit(1)
             }
 
-            printer.println(Ansi.ansi().fgGreen().a("TESTS PASSED").reset())
-        }
-    }
-
-    private class TestNameListener : TestExecutionListener {
-        override fun testPlanExecutionStarted(testPlan: TestPlan?) {
-            System.out.apply {
-                println()
-                println("----------------------------------")
-                println("SAML Conformance Test Kit Starting")
-                println("----------------------------------")
-            }
-        }
-
-        override fun executionFinished(
-            testIdentifier: TestIdentifier,
-            testExecutionResult: TestExecutionResult
-        ) {
-            if (testIdentifier.isTest)
-                System.out.println(
-                        """${testIdentifier.displayName}
-                            ${getResultDisplay(testExecutionResult.status)}""")
-        }
-
-        private fun getResultDisplay(status: TestExecutionResult.Status): Any {
-            return when (status) {
-                TestExecutionResult.Status.SUCCESSFUL ->
-                    Ansi.ansi().fgGreen().a(status.name).reset()
-                TestExecutionResult.Status.FAILED ->
-                    Ansi.ansi().fgRed().a(status.name).reset()
-                else -> status.name
```

Comment:
In order to preserve the information on the particular tests that failed, you could either:
1. Keep this `TestNameListener`, in addition to printing the `Report` at the end.
2. Somehow make either the `Report` or the `SAMLComplianceException`s aware of the current test being run.

Number 2 sounds really messy code-wise. There is an advantage to number 1, which is that it will show progress to the user as the tests are running. With this change as it is now, the user will not see any progress; they just get a summary at the end.

Reply:
And I'm realizing now that even that won't work as it is because the tests won't necessarily fail anymore... What if we gave the report some notion of "staged exceptions". In addition to preserving the exceptions it gets, it keeps a cache that the `TestNameListener` would grab at the end of the test and then clear out for the next test.